lib/sycamore/sycamore/writers/pinecone_writer.py

```diff
@@ -0,0 +1,207 @@
+            if len(prefix) == 0:
+                return str(key)
+            else:
+                return f"{prefix}.{key}"
+
+        def _flatten_metadata(data: Union[dict, list, tuple], prefix="") -> Iterable[Tuple[Any, Any]]:
+            # Pinecone requires metadata to be flat (no nested objects)
+            # so here's a traversal
+            iterator = []  # type: ignore
+            if isinstance(data, dict):
+                iterator = data.items()  # type: ignore
+            if isinstance(data, (list, tuple)):
+                iterator = enumerate(data)  # type: ignore
+            items = []
+            for k, v in iterator:
+                if isinstance(v, (dict, list, tuple)):
+                    if isinstance(v, (list, tuple)) and all(isinstance(innerv, str) for innerv in v):
+                        # Lists of strings are allowed
+                        items.append((_add_key_to_prefix(k, prefix), v))
+                    else:
+                        inner_values = _flatten_metadata(v, prefix=(_add_key_to_prefix(k, prefix)))
+                        items.extend([(innerk, innerv) for innerk, innerv in inner_values])
+                elif v is not None:
+                    items.append((_add_key_to_prefix(k, prefix), v))
+            return items
+
+        def _metadata_special_cases(data: dict) -> dict:
+            # Bytes are not allowed as pinecone metadata so we
+            # use a base64 encoded string
+            binary = data.pop("binary_representation", None)
+            if binary and self._include_binary:
+                b64binary = b64encode(binary)
+                strbinary = b64binary.decode("UTF-8")
+                data["binary_representation"] = strbinary
+            # We store shingles as a list of strings bc this is
+            # the only list type supported by pinecone metadata
+            # and it is sufficient for the purposes of NDD
+            shingles = data.get("shingles", None)
+            if shingles:
+                strshingles = [str(s) for s in shingles]
+                data["shingles"] = strshingles
+            # We store bboxes as named coordinates rather than
+            # just bbox.0, bbox.1, etc. Makes it cleaner
+            bbox = data.get("bbox", None)
+            if bbox:
+                bbox_as_dict = {
+                    "x1": bbox[0],
+                    "y1": bbox[1],
+                    "x2": bbox[2],
+                    "y2": bbox[3],
+                }
+                data["bbox"] = bbox_as_dict
+            # The rich Table class is obv out of bounds, so we
+            # store it as csv.
+            table = data.pop("table", None)
+            if table:
+                data["table"] = table.to_csv()
+            return data
+
+        def _record_to_object(record):
+            doc = Document.from_row(record)
+            data = doc.data
+            # Pull out pinecone top-level properties and leave
```

Comment:
The opensearch and weviate writers pull specific properties out of data and then pass through all of doc.properties.  I feel like we should do the same thing here (or plan to change the other two writers).

Reply:
basically what I'm doing here although I'll clean it up so it's a selection rather than a kitchen sink